Add experimental support for Apple VisionOS platform

komakai opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch

This is dependent on cmake support for visionos which is currently in progress.
Creating PR now to test that there are no regressions in iOS and macOS builds

komakai commented:
Build command for testing visionos support:

`./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs x86_64 --build_only_specified_archs`

asmorkalov commented on CMakeLists.txt:
```diff
@@ -462,6 +462,9 @@ OCV_OPTION(WITH_ANDROID_MEDIANDK "Use Android Media NDK for Video I/O (Android)"
 OCV_OPTION(WITH_ANDROID_NATIVE_CAMERA "Use Android NDK for Camera I/O (Android)" (ANDROID_NATIVE_API_LEVEL GREATER 23)
   VISIBLE_IF ANDROID
   VERIFY HAVE_ANDROID_NATIVE_CAMERA)
+OCV_OPTION(WITH_TENGINE "Include Arm Inference Tengine support" OFF
+  VISIBLE_IF (ARM OR AARCH64) AND (UNIX OR ANDROID) AND NOT IOS AND NOT VISIONOS
+  VERIFY HAVE_TENGINE)
```
TEngine support was dropped in some previous PRs as not supported by the original author. Please rebase on current 4.x.

komakai replied:
Rebased and force pushed

asmorkalov commented on platforms/ios/cmake/Toolchains/common-ios-toolchain.cmake:
```diff
@@ -158,7 +188,7 @@ else()
 
 # Include extra modules for the iOS platform files
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/platforms/ios/cmake/Modules")
+#set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/platforms/ios/cmake/Modules")
```
It breaks iOS configuration.

komakai replied:
I was testing to see if I need to add something similar for visionOS but commenting out this line didn't break anything on the iOS build so I'm not sure its needed any more. I can add it back just in case

komakai replied:
OK - its back now

asmorkalov commented:
@komakai Great job! Thanks a lot for the patch!

dkurt commented on CMakeLists.txt:
```diff
@@ -310,10 +310,10 @@ OCV_OPTION(WITH_WEBNN "Include WebNN support" OFF
   VERIFY HAVE_JASPER)
 OCV_OPTION(WITH_OPENJPEG "Include JPEG2K support (OpenJPEG)" ON
-  VISIBLE_IF NOT IOS
+  VISIBLE_IF NOT IOS AND NOT VISIONOS
```
I see a lot of `IOS AND VISIONOS` in pair. Is there a united variable for all? Like this:
```cmake

OCV_OPTION(WITH_OPENJPEG "Include JPEG2K support (OpenJPEG)" ON
  VISIBLE_IF NOT APPLE
```

komakai replied:
APPLE also includes macOS
I guess we could make something like APPLE_DEVICE but I feel like the names IOS and VISIONOS are going to be easier to understand for maintainers

komakai commented:
Converting to draft until the necessary changes to cmake are released

asmorkalov commented:
VisionOS added in CMake 3.28 release: https://cmake.org/cmake/help/latest/release/3.28.html
Related MR: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8581

asmorkalov commented:
VisionOS support is not complete in CMake for now: https://gitlab.kitware.com/cmake/cmake/-/issues/25266

komakai commented:
> VisionOS support is not complete in CMake for now: https://gitlab.kitware.com/cmake/cmake/-/issues/25266

I will try to build this PR with CMake-3.28.0-rc4 when I have a minute and check if it builds OK

komakai commented:
Could build with CMake from https://github.com/Kitware/CMake/releases/download/v3.28.0-rc4/cmake-3.28.0-rc4-macos-universal.dmg with the below change:
On iOS we have the following defines
*  **IPHONEOS** in **Toolchain-iPhoneOS_Xcode.cmake** (OpenCV file)
*  **IPHONESIMULATOR** in **Toolchain-iPhoneSimulator_Xcode.cmake**  (OpenCV file)
* **IOS** in **iOS-Initialize.cmake** (CMake file)

However there is no equivalent define in **visionOS-Initialize.cmake** so I added a define **XROS** to **Toolchain-XROS_Xcode.cmake** and **Toolchain-XRSimulator_Xcode.cmake** which is effectly equal to **VISIONOS** or **VISIONSIMULATOR** and updated all the cmake files to use this define instead of VISIONOS

komakai commented:
To test this I was able to create a VisionOS project and add the OpenCV framework and add some simple calls to create a **Mat** and do some operations on the **Mat**. I could see that the code built OK but I don't have a device and the latest VisionOS simulator only runs on Arm Silicon so I could not actually run the code.
Unless there is someone with an Arm Silicon Mac who can test this I suggest to accept this PR as is since I don't know when I'll get a chance to test it more thoroughly

gdyuldin commented:
Hi.
Tested under simulator on Apple Silicon (M1) with applying GaussianBlur - it works, when using framework from `visionsimulator`

Build command:
`./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs arm64 --build_only_specified_archs`

Tools versions:
* cmake - 3.28.0-rc4
* Xcode. -Version 15.1 beta (15C5028h)
* macOS - 13.5.1 (22G90) (Ventura)

With framework from `visionos` directory build failed with linker error: `ld: framework 'opencv2' not found`

fengyuentau commented:
Did I miss something to get a successful compile? I tested with:

- macOS Sonoma 14.1.1
- Xcode 15.0.1
- CMake 3.28.0-rc5 (cmake-3.28.0-rc5-macos-universal.tar.gz)

and both commands:

```
./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs arm64 --build_only_specified_archs

./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs x86_64 --build_only_specified_archs
```

but only got errors:

```
CMake Error at /workspace/cmake-3.28.0-rc5-macos-universal/CMake.app/Contents/share/cmake-3.28/Modules/Platform/visionOS-Initialize.cmake:4 (message):
  xros is not an visionOS SDK
Call Stack (most recent call first):
  /workspace/cmake-3.28.0-rc5-macos-universal/CMake.app/Contents/share/cmake-3.28/Modules/CMakeSystemSpecificInitialize.cmake:34 (include)
  CMakeLists.txt:117 (enable_language)
```

gdyuldin commented:
> Did I miss something to get a successful build? I tested with:
> 
> * macOS Sonoma 14.1.1
> * Xcode 15.0.1
> * CMake 3.28.0-rc5 (cmake-3.28.0-rc5-macos-universal.tar.gz)
> 
> and both commands:
> 
> ```
> ./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs arm64 --build_only_specified_archs
> 
> ./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs x86_64 --build_only_specified_archs
> ```
> 
> but only got errors:
> 
> ```
> CMake Error at /workspace/cmake-3.28.0-rc5-macos-universal/CMake.app/Contents/share/cmake-3.28/Modules/Platform/visionOS-Initialize.cmake:4 (message):
>   xros is not an visionOS SDK
> Call Stack (most recent call first):
>   /workspace/cmake-3.28.0-rc5-macos-universal/CMake.app/Contents/share/cmake-3.28/Modules/CMakeSystemSpecificInitialize.cmake:34 (include)
>   CMakeLists.txt:117 (enable_language)
> ```

Looks like xcode 15.0.1 doesn't support VisionOS ([link](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_0_1-release-notes))
You should use beta version of xcode (for example - [this](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_1-release-notes) )

fengyuentau commented:
> Looks like xcode 15.0.1 doesn't support VisionOS ([link](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_0_1-release-notes))
You should use beta version of xcode (for example - [this](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_1-release-notes) )

Ok, thanks. Let me try with xcode beta then.

komakai commented:
> Do you know how to configure CMake to use Xcode beta? I have installed Xcode 15.1 beta but CMake still complains `xros is not an visionOS SDK`.

Use **xcode-select** to select the beta-version - then CMake will use that version

fengyuentau commented:
> > Do you know how to configure CMake to use Xcode beta? I have installed Xcode 15.1 beta but CMake still complains `xros is not an visionOS SDK`.
> 
> Use **xcode-select** to select the beta-version - then CMake will use that version

Thank you for the instructions. I can confirm that build can be done successfully with cmake-3.28.0-rc5 and Xcode-15.1-beta and command `./opencv/platforms/apple/build_xcframework.py --out visionostest --visionos_archs arm64 --visionsimulator_archs arm64 --build_only_specified_archs`.

cc @asmorkalov

asmorkalov reviewed (approved):
:+1:

klatism commented:
@komakai I'm getting:

> /Users/klatism/Git/opencv/modules/videoio/src/cap_avfoundation.mm:1314:32: error: unrecognized platform name visionOS
>         if (@available(iOS 11, visionOS 1, *)) {

with:
MacOS: Ventura 13.6.1
XCode: 14.3.1
CMake: 3.27.0

that if statement seems to be within `#if TARGET_OS_VISION` block. I'm wondering could the check for visionOS be removed from @available?

komakai commented:
@klatism 
> I'm wondering could the check for visionOS be removed from @available?

Yes that sounds completely reasonable. I look forward to your Pull Request

klatism commented:
@komakai please take a look: https://github.com/opencv/opencv/pull/25189. Thanks!

komakai commented:
> @komakai please take a look: #25189. Thanks!

Thanks - looks good
